lib/sedna/algorithms/hard_example_mining/__init__.py

```diff
@@ -0,0 +1,150 @@
+
+    def __call__(self, infer_result=None):
+        """predict function, and it must be implemented by
+        different methods class.
+
+        :param infer_result: prediction result
+        :return: `True` means hard sample, `False` means not a hard sample.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def data_check(cls, data):
+        """Check the data in [0,1]."""
+        return 0 <= float(data) <= 1
+
+
+@ClassFactory.register(ClassType.HEM, alias="Threshold")
+class ThresholdFilter(BaseFilter, abc.ABC):
+    def __init__(self, threshold=0.5, **kwargs):
+        self.threshold = threshold
+
+    def __call__(self, infer_result=None):
+        """
+        :param infer_result: [N, 6], (x0, y0, x1, y1, score, class)
+        :return: `True` means hard sample, `False` means not a hard sample.
+        """
+        if not infer_result:
+            return True
+
+        image_score = 0
+        for bbox in infer_result:
+            image_score += bbox[4]
+
+        average_score = image_score / (len(infer_result) or 1)
+        return average_score < self.threshold
+
+
+@ClassFactory.register(ClassType.HEM, alias="CrossEntropy")
+class CrossEntropyFilter(BaseFilter, abc.ABC):
+    """ Implement the hard samples discovery methods named IBT
+    (image-box-thresholds).
+
+    :param threshold_cross_entropy: threshold_cross_entropy to filter img,
+                        whose hard coefficient is less than
+                        threshold_cross_entropy. And its default value is
+                        threshold_cross_entropy=0.5
+    """
+
+    def __init__(self, threshold_cross_entropy=0.5, **kwargs):
+        self.threshold_cross_entropy = threshold_cross_entropy
+
+    def __call__(self, infer_result=None):
+        """judge the img is hard sample or not.
+
+        :param infer_result:
+            prediction classes list,
+                such as [class1-score, class2-score, class2-score,....],
+                where class-score is the score corresponding to the class,
+                class-score value is in [0,1], who will be ignored if its value
+                 not in [0,1].
+        :return: `True` means a hard sample, `False` means not a hard sample.
+        """
+        if infer_result is None:
+            return False
+        elif len(infer_result) == 0:
+            return False
+        else:
+            log_sum = 0.0
+            data_check_list = [class_probability for class_probability
+                               in infer_result
+                               if self.data_check(class_probability)]
+            if len(data_check_list) == len(infer_result):
+                for class_data in data_check_list:
+                    log_sum += class_data * math.log(class_data)
+                confidence_score = 1 + 1.0 * log_sum / math.log(
+                    len(infer_result))
+                return confidence_score >= self.threshold_cross_entropy
```

Comment:
fix it done by https://github.com/kubeedge/sedna/pull/68

Reply:
fix